dockerclient/archive.go

```diff
@@ -182,9 +183,93 @@ func archiveFromDisk(directory string, src, dst string, allowDownload bool, excl
 		return nil, nil, err
 	}
 
-	klog.V(4).Infof("Tar of %s %#v", directory, options)
-	rc, err := archive.TarWithOptions(directory, options)
-	return rc, rc, err
+	pipeReader, pipeWriter := io.Pipe() // the archive we're creating
+
+	includeFiles := options.IncludeFiles
+	var returnedError error
+	go func() {
+		defer pipeWriter.Close()
+		tw := tar.NewWriter(pipeWriter)
+		defer tw.Close()
+		var nonArchives []string
+		for _, includeFile := range includeFiles {
+			if allowDownload && src != "." && src != "/" && isArchivePath(filepath.Join(directory, includeFile)) {
+				// it's an archive -> copy each item to the
+				// archive being written to the pipe writer
+				klog.V(4).Infof("Extracting %s", includeFile)
+				f, err := os.Open(filepath.Join(directory, includeFile))
+				if err != nil {
+					returnedError = err
+					return
+				}
+				dc, err := archive.DecompressStream(f)
+				if err != nil {
+					f.Close()
+					returnedError = err
+					return
+				}
+				tr := tar.NewReader(dc)
+				hdr, err := tr.Next()
+				for err == nil {
+					if renamed, ok := options.RebaseNames[includeFile]; ok {
+						hdr.Name = strings.TrimSuffix(renamed, includeFile) + hdr.Name
+						if hdr.Typeflag == tar.TypeLink {
+							hdr.Linkname = strings.TrimSuffix(renamed, includeFile) + hdr.Linkname
+						}
+					}
+					tw.WriteHeader(hdr)
+					_, err = io.Copy(tw, tr)
+					if err != nil {
+						break
```

Comment:
Do we care about this err?  Seems like it could just get lost?  Should it at least be logged somewhere?